### DISCUSSION TITLE
CEI smart contract design framework for `pickWinner` function in lesson 9

### DISCUSSION THREAD

**Discussion by Kuusho** [Cyfrin/foundry-full-course-cu#1180] [Q&A]
In lesson 9 of the foundry course, Patrick mentioned the CEI smart contract design framework. 

I'd like to ask, in the `pickWinner` function for the Raffle, 

`  
        uint256 requestId = i_vrfCoordinator.requestRandomWords(
            i_gasLane, i_subscriptionId, REQUEST_CONFRIMATIONS, i_callbackGasLimit, NUM_WORDS 
    }
`

 does this part that involves interacting with Chainlink  count as effects (since we imported the contracts) or are they interactions (because they are fetching data from chainlink)?

**Comment by xyz899** [comment]
### Hi !

The following line you wrote is an Interaction ( meaning we fetch data ) because you are asking of the Chainlink VRF to get random words that let's you generate a uniqueId. So no, in order to get a generated Id we need to fetch the data from an external source, like you do when fetching the price of a supported token with AggregatorV3Interface for example. 

_winter99_

**Reply by Kuusho** [reply]
Thanks for this answer! Absolutely well thought out, I appreciate it

**Comment by r0bchain** [comment]
Guys, I am trying to understand **where the call to the Chainlink VRF is made**. I see that the method is in an interface and does not implement the body of the method. Where is it implemented?
As I read, the implementation of the body is in **VRFConsumerBaseV2.sol**, but there is not this method, I only can see **fulfillRandomWords** but nothing about **requestRandomWords** , any clue?

```
  function requestRandomWords(
    bytes32 keyHash,
    uint64 subId,
    uint16 minimumRequestConfirmations,
    uint32 callbackGasLimit,
    uint32 numWords
  ) external returns (uint256 requestId);
```

**Reply by r0bchain** [reply]
Ok, I found the answer to my question here, I guess this is where the body of the method is located : https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol
I got it from the Chainlink VRF official documentation:
https://docs.chain.link/vrf/v2-5/overview/subscription

**Reply by EngrPips** [reply]
Yeah, You call the `VRFCoordinatorV2_5` to request random words, and the `VRFCoordinatorV2_5` will call your contract back, which must be a type of `VRFConsumerBaseV2` to fulfill your request for random words.
